[Processor] Implement eventProcessor (#3478)
nuclio/nuclio: docs/concepts/architecture.md

Reviewer: Great stuff! can you make sure this doesn't conflict with any of the other sections written above?
Also, I think it fits better in the `Processor architecture` section above.

    > **Note:** The shell runtime supports only file data bindings.

#### Processors internal abstractions

Here is an illustration of the abstraction levels that every event goes through while being processed:
![event-processing-flow](../../docs/assets/images/internal-processor-abstractions.png)

Thus, every event goes through two EventProcessor levels: the first at the trigger level and the second at the runtime level.
The EventProcessor interface generalizes the allocation logic, and shares it between different resources

The event processor allocation logic varies based on the expected behavior. Nuclio supports two modes:

- Synchronous (default) – Events within a single worker are processed sequentially (FIFO).
- Asynchronous – Events within a single worker are processed concurrently.

> Depending on the configured mode, Nuclio users different event processor allocators at different levels.

Reviewer: ```suggestion
Depending on the configured mode, Nuclio uses different event processor allocators at different levels.
```

The event-source mapping can either specify the exact function version, or utilize a defined alias and thus eliminate the need to change the mapping when a newer version of the function is published.

> ## Processors' internal abstractions

Reviewer: ```suggestion
## Processor internal abstractions
```

> ## Processors' internal abstractions

Reviewer 2: add a bullet at the top of the page for this section

Author: @jillnogold changed this chapter level to ####, so I believe not relevant anymore

Here is an illustration of the abstraction levels that every event goes through while being processed:
![event-processing-flow](../../docs/assets/images/internal-processor-abstractions.png)

Thus, every event goes through two EventProcessor levels: the first at the trigger level and the second at the runtime level.

> The EventProcessor exists to generalize the allocation logic and make it abstract.

Reviewer: ```suggestion
The purpose of the EventProcessor interface is to generalize the allocation logic, and share it between different resources.
```

> The EventProcessor exists to generalize the allocation logic and make it abstract.

Reviewer 2: The EventProcessor interface generalizes the allocation logic, and shares it between different resources.

> The event processor allocation logic varies based on the expected behavior. We support two modes:

Reviewer 2: The event processor allocation logic varies based on the expected behavior. Nuclio supports two modes:

> Synchronous (default) – Events within a single worker are processed sequentially (FIFO).
> Asynchronous – Events within a single worker are processed concurrently.

Reviewer: ```suggestion
- Synchronous (default) – Events within a single worker are processed sequentially (FIFO).
- Asynchronous – Events within a single worker are processed concurrently.
```

> Depending on the required behavior, we use different event processor allocators at different levels.

Reviewer: ```suggestion
Depending on the configured mode, we use different event processor allocators at different levels.
```

> Depending on the required behavior, we use different event processor allocators at different levels.

Reviewer 2: Depending on the configured mode, Nuclio users different event processor allocators at different levels.

> For instance, to support synchronous processing, we allocate a worker (event processor at the trigger level).

Reviewer 2: For instance, to support synchronous processing, Nuclio allocates a worker (event processor at the trigger level).

> While the allocated event processor is waiting for a response, it cannot be allocated again. To enforce this, we use a sync pool allocator,

Reviewer: ```suggestion
While the allocated worker is waiting for a response, it cannot be allocated again. To enforce this, we use a sync pool allocator,
```

> While the allocated event processor is waiting for a response, it cannot be allocated again. To enforce this, we use a sync pool allocator,

Reviewer 2: While the allocated worker is waiting for a response, it cannot be allocated again. Nuclio enforces this with a sync pool allocator that ensures that each event processor is assigned to only one event at a time.

> which ensures that each event processor is assigned to only one event at a time.

Reviewer: ```suggestion
which ensures that each worker is assigned to only one event at a time.
```

For asynchronous processing, a worker can be allocated for multiple events simultaneously.

> In this case, we use a singleton async allocator, which allows multiple events to share the same event processor without blocking further allocations.

Reviewer 2: At this stage, Nuclio uses a singleton async allocator, which allows multiple events to share the same event processor without blocking further allocations.

Author: @jillnogold here I meant In async processing case. Is it still ok to use `at the stage?`

> However, as the event moves further through the processing flow, we still need to allocate an event processor at the runtime level. While an event is being processed, we must ensure that no other events are processed through the same event processor to maintain correct behaviour.

Reviewer 2: This sentence isn't clear (to me). Do you mean:
However, as the event moves through the processing flow, we still need to allocate an event processor at the runtime level, it is allocated an event processor that is not used by any other events.

Author: @jillnogold not really, I meant that the event processor  Nuclio allocated on runtime level, may be used to process only one event at the same time. So nuclio ensures that no other events are processed through the same event processor to maintain correct behaviour.

Author: not sure how to make it better. Maybe smth like:

```
While an event is being processed by allocated event processor on runtime level, Nuclio ensures that no other events are processed through the same event processor to maintain correct behaviour.

```

> ### Synchronous vs. Asynchronous Allocator in Our Case

Reviewer 2: ### Synchronous vs. asynchronous allocator
